### ISSUE TITLE
[API Proposal]: PipeStream.WaitForPipeDrainAsync(CancellationToken cancellationToken)

### ISSUE THREAD

**Issue by Symbai** [dotnet/runtime#72593]
### Background and motivation

Currently `WaitForPipe()` is blocking the thread. Adding an async version with support for timeout let us better handle situations where the client cannot or has not read all remaining bytes within an expected duration. I.e. network failure or delay or client has been shut down etc.

### API Proposal

```csharp
namespace System.IO.Pipes

public abstract class PipeStream : Stream
{
    public Task WaitForPipeDrainAsync(CancellationToken cancellationToken);
}
```

### API Usage

```csharp
var ct = new CancellationTokenSource(3_000).Token;
try
{
    await NamedPipe.WaitForPipeDrainAsync(ct)
}
catch (OperationCanceledException e)
{
    //Retry or do something else
}



### Alternative Designs

_No response_

### Risks

_No response_

**Comment by Jozkee** [comment]
WaitForPipeDrain uses FlushFileBuffers, do you know which WinAPI should be used for this async version?

**Comment by Symbai** [comment]
No, I have no idea how it works internally. But since we also have async file IO APIs like `WriteAllBytesAsync` and there is no async windows API for that too, I would assume `WaitForPipeDrainAsync` does not really require an async windows API as well.

**Comment by stephentoub** [comment]
> since we also have async file IO APIs like WriteAllBytesAsync and there is no async windows API for that too

We do.  WriteAllBytesAsync is implemented on top of WriteFile, which supports overlapped I/O.

**Comment by Symbai** [comment]
I think I've seen existing .NET APIs which provide an async version which itself is just using a `Task` and wait for its result.

**Comment by stephentoub** [comment]
> I think I've seen existing .NET APIs which provide an async version which itself is just using a Task and wait for its result.

Only when we're forced to due to e.g. deriving from a base type that provides a virtual/abstract we need to override.

> How does FlushAsync work, is there an async windows API?

FileStream.FlushASync?  It takes the buffer stored in the FileStream and writes it out to the file essentially using WriteAsync.
